# Arrests Data Cleaning

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [2]:
# Load Data
df = pd.read_csv("data/arrests_adult-arrests-details_arrestdetail.csv")

The data was obtained from City of Phoenix Open Data at: https://www.phoenixopendata.com/dataset/arrests

In [3]:
# Print first 5 rows of data
df.head()

ARST_NUM  DATE_OCCUR DAY_OF_WEEK       MONTH QTR  YEAR ARREST_TYPE  \
0  PHX201801013548  01/01/2018    2-MONDAY  01-JANUARY  Q1  2018           O   
1  PHX201801013538  01/01/2018    2-MONDAY  01-JANUARY  Q1  2018           T   
2  PHX201801013560  01/01/2018    2-MONDAY  01-JANUARY  Q1  2018           O   
3  PHX201801013490  01/01/2018    2-MONDAY  01-JANUARY  Q1  2018           O   
4  PHX201801013488  01/01/2018    2-MONDAY  01-JANUARY  Q1  2018           T   

  UNIQUE_NAME_ID SUBJ_SEX                 SUBJ_RACE SUBJ_ETHNICITY  \
0    MNI-3059468     Male                     White       Hispanic   
1  MNI-100947779     Male                     White       Hispanic   
2    MNI-1270697     Male                     White   Non-Hispanic   
3  MNI-100947321     Male                     White       Hispanic   
4  MNI-100947309     Male  Asian / Pacific Islander   Non-Hispanic   

  SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP ARST_OFFICER ARST_OFFICER_SEX  \
0           Hispanic        39            30s        09284           Female   
1           Hispanic        23            20s        07558             Male   
2              White        35            30s        09764             Male   
3           Hispanic        40            40s        10044             Male   
4              Other        21            20s        08207             Male   

  ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP   HUNDREDBLOCKADDR  PRECINCT_NUM  \
0             White              White  35XX W DOBBINS RD         800.0   
1             Asian              Other   49XX W THOMAS RD         800.0   
2             White              White    96XX N 16TH AVE         600.0   
3          Hispanic           Hispanic   60XX W THOMAS RD         800.0   
4             White              White    34XX N 83RD AVE         800.0   

            PRECINCT  BEAT_NUM      BEAT MAPGRID  COUNCIL_DISTRICT_NUM  \
0  Maryvale/Estrella     834.0  834 Beat    IC20                   7.0   
1  Maryvale/Estrella     824.0  824 Beat    BE17                   4.0   
2     Desert Horizon     623.0  623 Beat    CH25                   3.0   
3  Maryvale/Estrella     821.0  821 Beat    BE14                   7.0   
4  Maryvale/Estrella     812.0  812 Beat    BG08                   5.0   

     COUNCIL_DISTRICT  FELONY_CHARGES  MISDEMEANOR_CHARGES  OTHER_CHARGES  \
0  Council District 7               0                    1              0   
1  Council District 4               6                    0              0   
2  Council District 3               0                    2              0   
3  Council District 7               0                    1              0   
4  Council District 5               1                    0              0   

   UNKNOWN_CHARGES  P1VIOLENT_CHARGES  P1PROPERTY_CHARGES  P2DRUG_CHARGES  \
0                0                  0                   0               0   
1                0                  0                   0               4   
2                0                  0                   0               0   
3                0                  1                   0               0   
4                0                  0                   0               0   

   ASSAULTOFFICER_CHARGES  RESISTARST_CHARGES  
0                       0                   0  
1                       0                   0  
2                       0                   0  
3                       0                   0  
4                       0                   0

In [4]:
# Print number of rows and columns
df.shape

(238890, 35)

In [5]:
# Print data types
df.dtypes

ARST_NUM                   object
DATE_OCCUR                 object
DAY_OF_WEEK                object
MONTH                      object
QTR                        object
YEAR                        int64
ARREST_TYPE                object
UNIQUE_NAME_ID             object
SUBJ_SEX                   object
SUBJ_RACE                  object
SUBJ_ETHNICITY             object
SIMPLE_SUBJ_RE_GRP         object
SUBJ_AGE                    int64
SUBJ_AGE_GROUP             object
ARST_OFFICER               object
ARST_OFFICER_SEX           object
ARST_OFFICER_RACE          object
SIMPLE_EMPL_RE_GRP         object
HUNDREDBLOCKADDR           object
PRECINCT_NUM              float64
PRECINCT                   object
BEAT_NUM                  float64
BEAT                       object
MAPGRID                    object
COUNCIL_DISTRICT_NUM      float64
COUNCIL_DISTRICT           object
FELONY_CHARGES              int64
MISDEMEANOR_CHARGES         int64
OTHER_CHARGES               int64
UNKNOWN_CHARGE

In [6]:
# Find number missing values for each column
df.isna().sum()

ARST_NUM                     0
DATE_OCCUR                   0
DAY_OF_WEEK                  0
MONTH                        0
QTR                          0
YEAR                         0
ARREST_TYPE                  0
UNIQUE_NAME_ID               0
SUBJ_SEX                     0
SUBJ_RACE                    0
SUBJ_ETHNICITY               0
SIMPLE_SUBJ_RE_GRP           0
SUBJ_AGE                     0
SUBJ_AGE_GROUP               0
ARST_OFFICER                 2
ARST_OFFICER_SEX             5
ARST_OFFICER_RACE            5
SIMPLE_EMPL_RE_GRP           5
HUNDREDBLOCKADDR          3069
PRECINCT_NUM              6953
PRECINCT                  6953
BEAT_NUM                  7038
BEAT                      7038
MAPGRID                   7022
COUNCIL_DISTRICT_NUM      7105
COUNCIL_DISTRICT             0
FELONY_CHARGES               0
MISDEMEANOR_CHARGES          0
OTHER_CHARGES                0
UNKNOWN_CHARGES              0
P1VIOLENT_CHARGES            0
P1PROPERTY_CHARGES           0
P2DRUG_C

In [7]:
# Drop Council District Number since it is part of the name in the Council District column
df.drop(['COUNCIL_DISTRICT_NUM'], axis=1)

ARST_NUM  DATE_OCCUR  DAY_OF_WEEK       MONTH QTR  YEAR  \
0       PHX201801013548  01/01/2018     2-MONDAY  01-JANUARY  Q1  2018   
1       PHX201801013538  01/01/2018     2-MONDAY  01-JANUARY  Q1  2018   
2       PHX201801013560  01/01/2018     2-MONDAY  01-JANUARY  Q1  2018   
3       PHX201801013490  01/01/2018     2-MONDAY  01-JANUARY  Q1  2018   
4       PHX201801013488  01/01/2018     2-MONDAY  01-JANUARY  Q1  2018   
...                 ...         ...          ...         ...  ..   ...   
238885  PHX202504308385  04/30/2025  4-WEDNESDAY    04-APRIL  Q2  2025   
238886  PHX202504308429  04/30/2025  4-WEDNESDAY    04-APRIL  Q2  2025   
238887  PHX202504308511  04/30/2025  4-WEDNESDAY    04-APRIL  Q2  2025   
238888  PHX202504308388  04/30/2025  4-WEDNESDAY    04-APRIL  Q2  2025   
238889  PHX202504308396  04/30/2025  4-WEDNESDAY    04-APRIL  Q2  2025   

       ARREST_TYPE UNIQUE_NAME_ID SUBJ_SEX                         SUBJ_RACE  \
0                O    MNI-3059468     Male                             White   
1                T  MNI-100947779     Male                             White   
2                O    MNI-1270697     Male                             White   
3                O  MNI-100947321     Male                             White   
4                T  MNI-100947309     Male          Asian / Pacific Islander   
...            ...            ...      ...                               ...   
238885           O   MNI-18139184     Male  American Indian / Alaskan Native   
238886           O    MNI-2924918     Male                             Black   
238887           T  MNI-103151855     Male                             Black   
238888           T  MNI-102696496     Male                             White   
238889           O  MNI-102381219   Female  American Indian / Alaskan Native   

       SUBJ_ETHNICITY SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP  \
0            Hispanic           Hispanic        39            30s   
1            Hispanic           Hispanic        23            20s   
2        Non-Hispanic              White        35            30s   
3            Hispanic           Hispanic        40            40s   
4        Non-Hispanic              Other        21            20s   
...               ...                ...       ...            ...   
238885   Non-Hispanic              Other        34            30s   
238886   Non-Hispanic              Black        36            30s   
238887   Non-Hispanic              Black        22            20s   
238888       Hispanic           Hispanic        29            20s   
238889   Non-Hispanic              Other        35            30s   

       ARST_OFFICER ARST_OFFICER_SEX ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP  \
0             09284           Female             White              White   
1             07558             Male             Asian              Other   
2             09764             Male             White              White   
3             10044             Male          Hispanic           Hispanic   
4             08207             Male             White              White   
...             ...              ...               ...                ...   
238885        10249             Male          Hispanic           Hispanic   
238886        06960             Male        OtherBlend              Other   
238887        11948             Male             White              White   
238888        11970             Male             White              White   
238889        09117             Male          Hispanic           Hispanic   

                    HUNDREDBLOCKADDR  PRECINCT_NUM           PRECINCT  \
0                  35XX W DOBBINS RD         800.0  Maryvale/Estrella   
1                   49XX W THOMAS RD         800.0  Maryvale/Estrella   
2                    96XX N 16TH AVE         600.0     Desert Horizon   
3                   60XX W THOMAS RD         800.0  Maryvale/Estrella   
4                    34XX N 83RD AVE         8

In [8]:
# Drop the numbers from Day of the Week and Month, leaving only the text
df['DAY_OF_WEEK'] = df['DAY_OF_WEEK'].str[2:]
df['MONTH'] = df['MONTH'].str[3:]

In [9]:
# CHeck all unique values for Month
df['MONTH'].unique()

array(['JANUARY', 'FEBRUARY', 'MARCH', 'APRIL', 'MAY', 'JUNE', 'JULY',
       'AUGUST', 'SEPTEMBER', 'OCTOBER', 'NOVEMBER', 'DECEMBER'],
      dtype=object)

In [10]:
# Check all unique values for Council District
df['COUNCIL_DISTRICT'].unique()

array(['Council District 7', 'Council District 4', 'Council District 3',
       'Council District 5', 'Council District 6', 'Council District 1',
       'Council District 8', 'Council District 2', 'Council District NA'],
      dtype=object)

In [11]:
# Convert the Council District NA to nulls
df.loc[df['COUNCIL_DISTRICT'].str.contains('Council District NA', na=False), 'COUNCIL_DISTRICT'] = np.nan

In [12]:
# Check unique age groups
df['SUBJ_AGE_GROUP'].unique()

array(['30s', '20s', '40s', '<20', '50s', '60s', '70s', '80s', '90s',
       '120s'], dtype=object)

In [13]:
# Check all ages above 90
df[df['SUBJ_AGE'] > 90]

ARST_NUM  DATE_OCCUR DAY_OF_WEEK     MONTH QTR  YEAR  \
227476  PHX202412027785  12/01/2024      SUNDAY  DECEMBER  Q4  2024   
232961  PHX202502168162  02/16/2025      SUNDAY  FEBRUARY  Q1  2025   

       ARREST_TYPE UNIQUE_NAME_ID SUBJ_SEX SUBJ_RACE SUBJ_ETHNICITY  \
227476           S  MNI-100952064     Male     White   Non-Hispanic   
232961           O  MNI-103152132     Male     White       Hispanic   

       SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP ARST_OFFICER  \
227476              White        92            90s        10910   
232961           Hispanic       125           120s        11125   

       ARST_OFFICER_SEX ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP  \
227476             Male             White              White   
232961             Male             White              White   

                 HUNDREDBLOCKADDR  PRECINCT_NUM        PRECINCT  BEAT_NUM  \
227476  W BECK LN & W GREENWAY RD         200.0  Black Mountain     221.0   
232961            154XX N 29TH ST         200.0  Black Mountain     232.0   

            BEAT MAPGRID  COUNCIL_DISTRICT_NUM    COUNCIL_DISTRICT  \
227476  221 Beat    DF25                   3.0  Council District 3   
232961  232 Beat    DF34                   2.0  Council District 2   

        FELONY_CHARGES  MISDEMEANOR_CHARGES  OTHER_CHARGES  UNKNOWN_CHARGES  \
227476               0                    1              0                0   
232961               0                    1              0                0   

        P1VIOLENT_CHARGES  P1PROPERTY_CHARGES  P2DRUG_CHARGES  \
227476                  0                   0               0   
232961                  1                   0               0   

        ASSAULTOFFICER_CHARGES  RESISTARST_CHARGES  
227476                       0                   0  
232961                       0                   0

The 125 year old is probably a typo and is most likely a 25 year old, given the violence charge. We will leave the 92 year old person, since the values are possible even if unlikely

In [14]:
# Example condition: ages above 100 are unrealistic
df.loc[df['SUBJ_AGE'] > 100, 'SUBJ_AGE'] = 25
df.loc[df['SUBJ_AGE_GROUP'].str.contains('120', na=False), 'SUBJ_AGE_GROUP'] = '20s'

In [15]:
# Check that the row has been corrected
df[df['SUBJ_AGE'] > 90]

ARST_NUM  DATE_OCCUR DAY_OF_WEEK     MONTH QTR  YEAR  \
227476  PHX202412027785  12/01/2024      SUNDAY  DECEMBER  Q4  2024   

       ARREST_TYPE UNIQUE_NAME_ID SUBJ_SEX SUBJ_RACE SUBJ_ETHNICITY  \
227476           S  MNI-100952064     Male     White   Non-Hispanic   

       SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP ARST_OFFICER  \
227476              White        92            90s        10910   

       ARST_OFFICER_SEX ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP  \
227476             Male             White              White   

                 HUNDREDBLOCKADDR  PRECINCT_NUM        PRECINCT  BEAT_NUM  \
227476  W BECK LN & W GREENWAY RD         200.0  Black Mountain     221.0   

            BEAT MAPGRID  COUNCIL_DISTRICT_NUM    COUNCIL_DISTRICT  \
227476  221 Beat    DF25                   3.0  Council District 3   

        FELONY_CHARGES  MISDEMEANOR_CHARGES  OTHER_CHARGES  UNKNOWN_CHARGES  \
227476               0                    1              0                0   

        P1VIOLENT_CHARGES  P1PROPERTY_CHARGES  P2DRUG_CHARGES  \
227476                  0                   0               0   

        ASSAULTOFFICER_CHARGES  RESISTARST_CHARGES  
227476                       0                   0

In [16]:
# Check that the row has been corrected
df.loc[df['ARST_NUM'] == 'PHX202502168162']

ARST_NUM  DATE_OCCUR DAY_OF_WEEK     MONTH QTR  YEAR  \
232961  PHX202502168162  02/16/2025      SUNDAY  FEBRUARY  Q1  2025   

       ARREST_TYPE UNIQUE_NAME_ID SUBJ_SEX SUBJ_RACE SUBJ_ETHNICITY  \
232961           O  MNI-103152132     Male     White       Hispanic   

       SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP ARST_OFFICER  \
232961           Hispanic        25            20s        11125   

       ARST_OFFICER_SEX ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP HUNDREDBLOCKADDR  \
232961             Male             White              White  154XX N 29TH ST   

        PRECINCT_NUM        PRECINCT  BEAT_NUM      BEAT MAPGRID  \
232961         200.0  Black Mountain     232.0  232 Beat    DF34   

        COUNCIL_DISTRICT_NUM    COUNCIL_DISTRICT  FELONY_CHARGES  \
232961                   2.0  Council District 2               0   

        MISDEMEANOR_CHARGES  OTHER_CHARGES  UNKNOWN_CHARGES  \
232961                    1              0                0   

        P1VIOLENT_CHARGES  P1PROPERTY_CHARGES  P2DRUG_CHARGES  \
232961                  1                   0               0   

        ASSAULTOFFICER_CHARGES  RESISTARST_CHARGES  
232961                       0                   0

In [17]:
# Change the data types of Precinct and Beat Number
df['PRECINCT_NUM'] = pd.to_numeric(df['PRECINCT_NUM'], errors='coerce').astype('Int64')
df['BEAT_NUM'] = pd.to_numeric(df['BEAT_NUM'], errors='coerce').astype('Int64')

In [18]:
# Check the data types again
df.dtypes

ARST_NUM                   object
DATE_OCCUR                 object
DAY_OF_WEEK                object
MONTH                      object
QTR                        object
YEAR                        int64
ARREST_TYPE                object
UNIQUE_NAME_ID             object
SUBJ_SEX                   object
SUBJ_RACE                  object
SUBJ_ETHNICITY             object
SIMPLE_SUBJ_RE_GRP         object
SUBJ_AGE                    int64
SUBJ_AGE_GROUP             object
ARST_OFFICER               object
ARST_OFFICER_SEX           object
ARST_OFFICER_RACE          object
SIMPLE_EMPL_RE_GRP         object
HUNDREDBLOCKADDR           object
PRECINCT_NUM                Int64
PRECINCT                   object
BEAT_NUM                    Int64
BEAT                       object
MAPGRID                    object
COUNCIL_DISTRICT_NUM      float64
COUNCIL_DISTRICT           object
FELONY_CHARGES              int64
MISDEMEANOR_CHARGES         int64
OTHER_CHARGES               int64
UNKNOWN_CHARGE

In [19]:
# Print the first 5 rows of the data
df.head()

ARST_NUM  DATE_OCCUR DAY_OF_WEEK    MONTH QTR  YEAR ARREST_TYPE  \
0  PHX201801013548  01/01/2018      MONDAY  JANUARY  Q1  2018           O   
1  PHX201801013538  01/01/2018      MONDAY  JANUARY  Q1  2018           T   
2  PHX201801013560  01/01/2018      MONDAY  JANUARY  Q1  2018           O   
3  PHX201801013490  01/01/2018      MONDAY  JANUARY  Q1  2018           O   
4  PHX201801013488  01/01/2018      MONDAY  JANUARY  Q1  2018           T   

  UNIQUE_NAME_ID SUBJ_SEX                 SUBJ_RACE SUBJ_ETHNICITY  \
0    MNI-3059468     Male                     White       Hispanic   
1  MNI-100947779     Male                     White       Hispanic   
2    MNI-1270697     Male                     White   Non-Hispanic   
3  MNI-100947321     Male                     White       Hispanic   
4  MNI-100947309     Male  Asian / Pacific Islander   Non-Hispanic   

  SIMPLE_SUBJ_RE_GRP  SUBJ_AGE SUBJ_AGE_GROUP ARST_OFFICER ARST_OFFICER_SEX  \
0           Hispanic        39            30s        09284           Female   
1           Hispanic        23            20s        07558             Male   
2              White        35            30s        09764             Male   
3           Hispanic        40            40s        10044             Male   
4              Other        21            20s        08207             Male   

  ARST_OFFICER_RACE SIMPLE_EMPL_RE_GRP   HUNDREDBLOCKADDR  PRECINCT_NUM  \
0             White              White  35XX W DOBBINS RD           800   
1             Asian              Other   49XX W THOMAS RD           800   
2             White              White    96XX N 16TH AVE           600   
3          Hispanic           Hispanic   60XX W THOMAS RD           800   
4             White              White    34XX N 83RD AVE           800   

            PRECINCT  BEAT_NUM      BEAT MAPGRID  COUNCIL_DISTRICT_NUM  \
0  Maryvale/Estrella       834  834 Beat    IC20                   7.0   
1  Maryvale/Estrella       824  824 Beat    BE17                   4.0   
2     Desert Horizon       623  623 Beat    CH25                   3.0   
3  Maryvale/Estrella       821  821 Beat    BE14                   7.0   
4  Maryvale/Estrella       812  812 Beat    BG08                   5.0   

     COUNCIL_DISTRICT  FELONY_CHARGES  MISDEMEANOR_CHARGES  OTHER_CHARGES  \
0  Council District 7               0                    1              0   
1  Council District 4               6                    0              0   
2  Council District 3               0                    2              0   
3  Council District 7               0                    1              0   
4  Council District 5               1                    0              0   

   UNKNOWN_CHARGES  P1VIOLENT_CHARGES  P1PROPERTY_CHARGES  P2DRUG_CHARGES  \
0                0                  0                   0               0   
1                0                  0                   0               4   
2                0                  0                   0               0   
3                0                  1                   0               0   
4                0                  0                   0               0   

   ASSAULTOFFICER_CHARGES  RESISTARST_CHARGES  
0                       0                   0  
1                       0                   0  
2                       0                   0  
3                       0                   0  
4                       0                   0

In [20]:
# Save the dataframe as a csv file for Tableau Visualization
df.to_csv('cleaned_arrest_data.csv', index=False, encoding='utf-8')